aws_lambda_powertools/utilities/idempotency/persistence/base.py

```diff
@@ -0,0 +1,432 @@
+        del self._cache[idempotency_key]
+
+    def save_success(self, event: Dict[str, Any], result: dict) -> None:
+        """
+        Save record of function's execution completing succesfully
+
+        Parameters
+        ----------
+        event: Dict[str, Any]
+            Lambda event
+        result: dict
+            The response from lambda handler
+        """
+        response_data = json.dumps(result)
+
+        data_record = DataRecord(
+            idempotency_key=self._get_hashed_idempotency_key(event),
+            status=STATUS_CONSTANTS["COMPLETED"],
+            expiry_timestamp=self._get_expiry_timestamp(),
+            response_data=response_data,
+            payload_hash=self._get_hashed_payload(event),
+        )
+        logger.debug(
+            f"Lambda successfully executed. Saving record to persistence store with "
+            f"idempotency key: {data_record.idempotency_key}"
+        )
+        self._update_record(data_record=data_record)
+
+        if self.use_local_cache:
+            self._save_to_cache(data_record)
+
+    def save_inprogress(self, event: Dict[str, Any]) -> None:
+        """
+        Save record of function's execution being in progress
+
+        Parameters
+        ----------
+        event: Dict[str, Any]
+            Lambda event
+        """
+        data_record = DataRecord(
+            idempotency_key=self._get_hashed_idempotency_key(event),
+            status=STATUS_CONSTANTS["INPROGRESS"],
+            expiry_timestamp=self._get_expiry_timestamp(),
+            payload_hash=self._get_hashed_payload(event),
+        )
+
+        logger.debug(f"Saving in progress record for idempotency key: {data_record.idempotency_key}")
+
+        if self.use_local_cache:
+            record = self._retrieve_from_cache(idempotency_key=data_record.idempotency_key)
+            if record:
+                raise ItemAlreadyExistsError
+
+        self._put_record(data_record)
+
+        # This has to come after _put_record. If _put_record call raises ItemAlreadyExists we shouldn't populate the
+        # cache with an "INPROGRESS" record as we don't know the status in the data store at this point.
+        if self.use_local_cache:
+            self._save_to_cache(data_record)
+
+    def save_error(self, event: Dict[str, Any], exception: Exception):
+        """
+        Save record of lambda handler raising an exception
+
+        Parameters
+        ----------
+        event: Dict[str, Any]
+            Lambda event
+        exception
+            The exception raised by the lambda handler
+        """
+        data_record = DataRecord(idempotency_key=self._get_hashed_idempotency_key(event))
+
+        logger.debug(
+            f"Lambda raised an exception ({type(exception).__name__}). Clearing in progress record in persistence "
+            f"store for idempotency key: {data_record.idempotency_key}"
+        )
+        self._delete_record(data_record)
```

Comment:
@cakepietoast  - i thought you wanted to save the record as completed, but with an error? 

This would mean any retries with the same idempotent key would not execute the lambda again.

Reply:
That was how the initial implementation looked, but there were too many gotchas to work around. We'd have to break the contract of "an invocation with the same payload (idempotency key) gets the same response" which means complicating customers' client implementations or orchestration logic.

The solution as its implemented now is simple. If the lambda handler raises an unhandled exception, we will allow it to be invoked again without "protecting" it with idempotency logic. This happens because we delete the idempotency record for that key. If you don't want the handler to be called again on a retry, you can catch the exception in your lambda handler and return a response instead of raising an exception.